### Demo
This is a demo notebook that loads a model from a checkpoint or pickle file through the `load_model` function and uses it to make predictions on a test set. A call is going to be made to the LLM to get recommendations based on the model's predictions.

In [2]:
from data_cleanup import DataProcessor
from model_loader import ModelLoader
from llm_recommendations import get_recommendations

In [ ]:
# load data
INPUT_STEPS = 24*3
OUTPUT_STEPS = 24

processor = DataProcessor(input_steps=INPUT_STEPS, output_steps=OUTPUT_STEPS)
(X_train, y_train), (X_val, y_val), (X_test, y_test) = processor.load_and_process_data()

Step 1/5: Fetching, cleaning, and engineering features...


/Users/felipejaracaceres/Documents/UCD/Machine Learning/ECS171G13/venv/lib/python3.11/site-packages/ucimlrepo/fetch.py:97: DtypeWarning: Columns (2,3,4,5,6,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(data_url)
/Users/felipejaracaceres/Documents/UCD/Machine Learning/ECS171G13/data_cleanup.py:135: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df = df.fillna(method='ffill')


Step 2/5: Resampling data to hourly and setting 'Global_active_power' as target...


/Users/felipejaracaceres/Documents/UCD/Machine Learning/ECS171G13/data_cleanup.py:175: FutureWarning: 'H' is deprecated and will be removed in a future version, please use 'h' instead.
  df_hourly = df.resample('H').agg(agg_dict)


Step 3/5: Splitting data and applying scaler...
Step 4/5: Creating time-series windows...
Step 5/5: Data processing complete.


/Users/felipejaracaceres/Documents/UCD/Machine Learning/ECS171G13/data_cleanup.py:176: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_hourly = df_hourly.fillna(method='ffill')


In [ ]:
# get the 4 best channels from the data: 
channel_indices = [4,5,6,7]
n_features = X_train.shape[2]
assert max(channel_indices) < n_features, 'Channel indices out of range for X_train features'

X_train_ch = X_train[:, :, channel_indices]  # shape (n_samples, time, 4)
X_val_ch   = X_val[:, :, channel_indices]
X_test_ch  = X_test[:, :, channel_indices]

print('Selected channel shapes:')
print('X_train_ch', X_train_ch.shape, 'y_train', y_train.shape)

Selected channel shapes:
X_train_ch (25832, 72, 4) y_train (25832, 24)


In [ ]:
# initialize model
model = ModelLoader("CNN")

predictions = model.get_predictions(X_test_ch[0])
print(predictions)

get_recommendations(predictions)

[0.0708092  0.06133023 0.04943363 0.05237386 0.04737218 0.06063753
 0.10861931 0.15706429 0.19417384 0.18936378 0.16729927 0.14631648
 0.12340467 0.11308512 0.10427341 0.08769623 0.07750625 0.10410301
 0.12693095 0.17431037 0.2182869  0.20802939 0.16941914 0.12493579]


'Based on the provided time series data of power consumption, we can analyze the trends, identify spikes and drops, and hypothesize potential causes for these fluctuations. Here’s a breakdown of the data and recommendations for reducing power consumption:\n\n### Data Analysis\n\n1. **Initial Observations**:\n   - The data shows a general trend of increasing power consumption from the beginning to the middle of the series, with a peak around the 20th data point (0.2182869).\n   - After the peak, there is a noticeable drop in consumption, followed by a slight recovery towards the end of the series.\n\n2. **Identifying Spikes and Drops**:\n   - **Spikes**: \n     - The most significant spike occurs between the 18th and 20th data points, where consumption rises from 0.12693095 to 0.2182869.\n   - **Drops**:\n     - A drop is observed after the peak, where consumption decreases from 0.2182869 to 0.12493579.\n\n### Hypothesized Causes\n\n1. **Spikes**:\n   - **Increased Activity**: The spike

#### Output

Based on the provided time series data of power consumption, we can analyze the trends, identify spikes and drops, and hypothesize potential causes for these fluctuations. Here’s a breakdown of the data and recommendations for reducing power consumption:\n\n### Data Analysis\n\n1. **Initial Observations**:\n   - The data shows a general trend of increasing power consumption from the beginning to the middle of the series, with a peak around the 20th data point (0.2182869).\n   - After the peak, there is a noticeable drop in consumption, followed by a slight recovery towards the end of the series.\n\n2. **Identifying Spikes and Drops**:\n   - **Spikes**: \n     - The most significant spike occurs between the 18th and 20th data points, where consumption rises from 0.12693095 to 0.2182869.\n   - **Drops**:\n     - A drop is observed after the peak, where consumption decreases from 0.2182869 to 0.12493579.\n\n### Hypothesized Causes\n\n1. **Spikes**:\n   - **Increased Activity**: The spike in power consumption could be attributed to increased activity in the household or facility, such as hosting events, using multiple appliances simultaneously (e.g., heating/cooling systems, kitchen appliances), or seasonal changes (e.g., increased heating in winter).\n   - **Equipment Malfunction**: A malfunctioning appliance may draw more power than usual, leading to a spike.\n\n2. **Drops**:\n   - **Reduced Activity**: The drop in consumption could indicate a period of reduced activity, such as weekends or holidays when fewer appliances are in use.\n   - **Energy-Saving Measures**: Implementation of energy-saving practices or the use of energy-efficient appliances could also contribute to the decrease in consumption.\n\n### Recommendations\n\n1. **Monitor Appliance Usage**:\n   - Keep track of which appliances are used during peak consumption times. Consider scheduling high-energy tasks (like laundry or dishwashing) during off-peak hours to balance the load.\n\n2. **Energy Audit**:\n   - Conduct an energy audit to identify appliances that consume excessive power. Replace or upgrade old appliances with energy-efficient models.\n\n3. **Smart Home Technology**:\n   - Invest in smart home technology that allows for better monitoring and control of energy consumption. Smart thermostats, for example, can optimize heating and cooling schedules.\n\n4. **Behavioral Changes**:\n   - Encourage household members to adopt energy-saving habits, such as turning off lights when not in use, unplugging devices, and using natural light whenever possible.\n\n5. **Regular Maintenance**:\n   - Ensure regular maintenance of heating and cooling systems to prevent inefficiencies that could lead to spikes in power consumption.\n\n6. **Consider Renewable Energy Sources**:\n   - If feasible, consider investing in renewable energy sources such as solar panels to offset power consumption from the grid.\n\nBy implementing these recommendations, users can effectively manage their power consumption, reduce spikes, and maintain a more consistent and lower energy usage profile.'